### ISSUE TITLE
Prevent mistakes by warning when LINQ method return value is not used, e.g. `list.Append(item);`

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn-analyzers#3363]
I think there are similar analyzers that warn if you don't use the return value of methods on immutable collections and Roslyn's immutable objects.

Something like this?

```cs
var list = new List<int>();

// ⚠ This method returns a new sequence which is not used. Did you mean 'list.Add'?
//   ↓↓↓↓↓↓↓↓↓↓
list.Append(42);
```

This should also apply to all LINQ methods known to be pure, like `Enumerable.Where`.

**Comment by Evangelink** [comment]
The rule is [CA1806](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1806?view=vs-2019).

I would have been expecting all of the linq methods to have the `PureAttribute` and so be handled by this rule already but that's not the case. I have opened a ticket on `dotnet/runtime`.

In the meantime, I think that would be nice to tweak the code to handle all methods from `System.Linq.Enumerable`.

@mavasani Do you think that's also worth adding a user option to provide more methods?

**Comment by stephentoub** [comment]
cc: @terrajobst

**Comment by mavasani** [comment]
> In the meantime, I think that would be nice to tweak the code to handle all methods from System.Linq.Enumerable.
@mavasani Do you think that's also worth adding a user option to provide more methods?

Yes and yes 👍

**Comment by JeremyMorton** [comment]
I am confused as to why this new warning is extended to Linq methods that take user-defined code as a Func or Action.

Since user-defined code can, and where we are seeing this pop-up, _does_ have side effects, it seems odd for the analyzer to warn and say, "Linq methods are known to not have side effects."

**Comment by jnm2** [comment]
Aren't side effects a code smell though? I never thought they went well with LINQ methods.

**Comment by Evangelink** [comment]
I am not a big fan of them but I do think that this rule should be tweaked to avoid reporting false-positives. It might be interesting to avoid a separate rule preventing changing a captured variable in a lambda but that's probably another thread/discussion. 

Getting back to the side-effect part, it's gonna be hard to detect them. We could stop reporting if there is any field on the left part of an assignment but this won't prevent cases where there is a method called (method causing a side-effect) nor cases where there is a capture with an assignment... @mavasani WDYT?

**Comment by JeremyMorton** [comment]
I agree that the side effects are a code smell, and appreciate a warning. My issue is that I think saying that "`System.Linq.Enumerable.First<TSource>(this IEnumerable<TSource> source)` is known to not cause side effects" is accurate, but saying "`System.Linq.Enumerable.First<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)` is known to not cause side effects" is _not_ accurate.

The specific case I'm seeing is using `System.Linq.Enumerable.Aggregate` to tack items onto a linked list, with the tail node being put into the accumulator each time. For the first couple of times, the return value is assigned to a variable that is passed as the initial value for the next call of `Aggregate`'s accumulator, but the last call doesn't need to store the final value of the tail node, so this new warning heavily implies that the last call has no effect and doesn't need to exist if it's return value is ignored.
